[SILGen] Generate a trap for unexpected cases in all @objc enums

jrose-apple opened the pull request:
(both C enums and Swift enums declared `@objc`), because of the "feature" in C of treating a value not declared as a case as a valid value of an enum.  No more undefined behavior here!

This bit can go in separately from all the work on exhaustive/frozen enums, which is still being discussed and will come later.

rdar://problem/20420436

jckarter commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -2600,17 +2549,21 @@ void SILGenFunction::emitSwitchStmt(SwitchStmt *S) {
+    // If we fail to match anything, we trap. This can happen with a switch
+    // over an @objc enum, which may contain any value of its underlying type.
+    // FIXME: Even if we can't say what the invalid value was, we should at
+    // least mention that this was because of a non-exhaustive enum.
```
I think it's OK to do it piecemeal.

jckarter commented:
Looks good. Do you have any tests for trapping on `@objc` enums in nested pattern positions?

vedantk commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -1679,113 +1679,143 @@ void PatternMatchEmission::emitIsDispatch(ArrayRef<RowToSpecialize> rows,
   SILValue srcValue = src.getFinalManagedValue().copy(SGF, loc).forward(SGF);
+  // FIXME: Pass caseCounts in here as well, as it is in
+  // emitEnumElementDispatch.
   SGF.B.createSwitchEnum(loc, srcValue, defaultBB, caseBBs);
```
Switch and case statements are handled for coverage reporting purposes. This might be more up @shajrawi's alley.

jrose-apple commented:
Ooh, good call. I'll add one now.

jrose-apple committed:
[SILGen] Generate a trap for unexpected cases in all @objc enums

(both C enums and Swift enums declared @objc), because of the
"feature" in C of treating a value not declared as a case as a valid
value of an enum.  No more undefined behavior here!

This bit can go in separately from all the work on exhaustive/frozen
enums, which is still being discussed and will come later.

rdar://problem/20420436

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Simple code size test:

@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test compiler performance

gottesmm commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -1679,113 +1679,143 @@ void PatternMatchEmission::emitIsDispatch(ArrayRef<RowToSpecialize> rows,
-  SILBasicBlock *defaultBB = nullptr;
+/// Create destination blocks for switching over the cases in an enum defined
+/// by \p rows.
+static void generateEnumCaseBlocks(
```
Two things:

1. Can you put the refactoring in a separate commit? That will make the diff even /easier/ to read than whitespace-insensitive diff.
2. I purposely duplicated the code so I could easily visualize the difference in between the two.

jrose-apple replied:
1. Fair.
2. Please don't do this in the future. You saw below that there was already an inconsistency in code coverage.

gottesmm replied:
@jrose-apple on 2:

1. At the time, it was going to be temporary, but other things took priority. When you are trying to make a large transition in a code base (like putting ownership into SILGen), it is "cleaner" to duplicate old -> new/fix new/remove the old. I was just interrupted by the +0 work. Once +0 is done, this is next on my list to fix.

2. I disagree strongly with your characterization that this code duplication directly resulted in the code coverage inconsistency. What resulted in the code coverage inconsistency is a lack of enough test coverage of our emission of code coverage. If we had test coverage of such paths, then the error would have never happened. That is like blaming a night time driver for hitting a hole dug by road construction crews, when the road construction crews did not put up any signs with warnings.

*NOTE* I am taking your word for the error. I have not looked at it.

gottesmm replied:
That being said, I do not think arguing about blame/etc about the coverage is productive, so lets drop that for now? ; ).

gottesmm commented on lib/SILGen/SILGenPattern.cpp:
```diff
@@ -1679,113 +1679,143 @@ void PatternMatchEmission::emitIsDispatch(ArrayRef<RowToSpecialize> rows,
-      } else {
-        specRow.Patterns.push_back(nullptr);
-      }
+  llvm::DenseMap<EnumElementDecl*, unsigned> caseToIndex;
```
Can you change this to be a SmallDenseMap? DenseMap seems really heavy weight here.

jrose-apple replied:
This isn't a change from what was there before, but sure. (Makes sense, most enums are small.)

gottesmm replied:
Yes. You are correct.

gottesmm reviewed (commented):
Can you move the refactoring into a separate commit/pr? Also a quick comment about a DenseMap.

jrose-apple committed:
[SILGen] Factor out the generation of blocks for switching on an enum

This code is shared between emitEnumElementDispatch and
emitEnumElementDispatchWithOwnership. Eventually the former will go
away, but for now there's no need to have both copies (which really
are the same).

No intended functionality change.

jrose-apple commented:
Already passed full-testing previously, so just doing a smoke test now.

@swift-ci Please smoke test

jrose-apple commented:
Filed rdar://problem/37728359 to improve the error message and assigned it to myself.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple committed:
[SILGen] Minor improvements to generateEnumCaseBlocks

Noticed by MichaelG while reviewing my changes in the previous commit.
No intended functionality change.

jrose-apple commented:
@swift-ci Please smoke test
